modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,343 @@
+package proto
+
+import (
+	"fmt"
+	"io"
+	"os"
+
+	"github.com/NebulousLabs/Sia/build"
+	"github.com/NebulousLabs/Sia/crypto"
+	"github.com/NebulousLabs/errors"
+)
+
+// merkleRootsCacheHeight is the height of the subTrees in cachedSubTrees. A
+// height of 7 means that 128 sector roots are covered by a single cached
+// subTree.
+const merkleRootsCacheHeight = 7
+
+type (
+	// merkleRoots is a helper struct that makes it easier to add/insert/remove
+	// merkleRoots within a SafeContract.
+	merkleRoots struct {
+		// cachedSubTrees are cached trees that can be used to more efficiently
+		// compute the merkle root of a contract. The cached trees are not
+		// persisted and are computed after startup.
+		cachedSubTrees []*cachedSubTree
+		// uncachedRoots contains the sector roots that are not part of a
+		// cached subTree. The uncachedRoots slice should never get longer than
+		// 2^merkleRootsCacheHeight since that would simply result in a new
+		// cached subTree in cachedSubTrees.
+		uncachedRoots []crypto.Hash
+
+		// file is the file of the safe contract that contains the root.
+		file *os.File
+		// numMerkleRoots is the number of merkle roots in file.
+		numMerkleRoots int
+	}
+
+	cachedSubTree struct {
+		height int
+		sum    crypto.Hash
+	}
+)
+
+// loadExistingMerkleRoots reads creates a merkleRoots object from existing
+// merkle roots.
+func loadExistingMerkleRoots(file *os.File) (mr *merkleRoots, err error) {
+	mr = &merkleRoots{
+		file: file,
+	}
+	// Get the number of roots stored in the file.
+	mr.numMerkleRoots, err = mr.lenFromFile()
+	if err != nil {
+		return
```

Comment:
convention would be to return `nil, err` anytime that there's an error, same follows for the rest of the function. For a function this long, it's we tend to strongly prefer unnamed return values, so that we have to explicitly state what is being returned.